### ISSUE TITLE
Cracking ZIP containing Deflate JPEGs when original decompressed files are known

### ISSUE THREAD

**Issue by kast1450** [kimci86/bkcrack#50]
Hello, I'm trying to crack a ZIP file containing numerous JPEGs.
I have a copy of one of the original JPEGs, CRC matching.
However, the JPEGs inside have been compressed with Deflate (zipdetails says Maximum).

I've tried all the compression tools for deflate however I've had no luck in cracking the ZIP.
Any advice?

**Comment by kimci86** [comment]
Hello!
You have one original file, so it should be possible to crack the archive. As you understood, you have to compress the know file with the same compression which was used in the encrypted archive.
As you do not know which tool or settings where used, you can try many. Compression tools usually have a compression level setting. You can try to tweak that.
A hint to know if the compression is the right one is to look at the compressed size. Knowing that encryption adds a 12 bytes header after compression, the goal is to have: compressed size = packed size in encrypted archive - 12.

If the attack still fails when the compressed size is good, you can try using a prefix of the compressed data. Indeed, we can hope that compression behaves like in the encrypted archive at the beginning, but diverges at some point. The command line option `-t` can be used to to so.
Let me know how it goes.

**Comment by kast1450** [comment]
Hello, I have tried what you suggested. The author claims the file was created using 7-Zip, and I was unable to match the exact file size however I was able to get within a few tens of bytes. I tested if this made enough of a difference by changing the compression settings and viewing it in a hex editor however there were next to no differences in the file.

However, using the command `bkcrack.exe -C cipher.zip -c image.jpg -P recreated.zip -p image.jpg -t 64` results in a "Could not find the keys." error.

**Comment by kimci86** [comment]
> It also appears to be getting the encrypted file offset wrong, as it says Attack on ... Z values at index 7 where as the encrypted plaintext is stored at offset 4,881,988,972.

The index 7 is relative to the beginning of ciphertext, so it looks good. Nothing to worry about here.

Besides, I am not sure what is going wrong.
Would you be ok with sharing a part of your encrypted archive and the original uncompressed file so I and others can have a try? You can drag and drop files here to upload them.
Before uploading, you should make a copy of the encrypted archive and remove entries with an archive manager to leave just the one corresponding to the known file, to save time and storage as your archive seems pretty big.
You can send files to me via email if you prefer not to share them with the entire internet. (Check the git log for my email)

**Comment by kast1450** [comment]
Thanks to the incredible support by @kimci86 they were able to successfully find the encryption keys for the archive using only one original JPEG as the plaintext.

Here's the final result for the ZIP file output by using the -U switch to generate an encrypted archive with a known password.
![Final Result](https://i.imgur.com/PQUg059.png)

I'm sure @kimci86 will write a better response, but for now here's the contents of the email sent to me. (I only modified the keys to hide possibly private info.)

~~~~
So I tried several compression tools (Info-ZIP, 7-Zip, WinZip) with
several compression levels and looked at the generated deflate data.
Indeed, I could not manage to get the expected compressed size. Looking
at the beginning of compressed data, there was some differences between
tools and compression levels, so not sure what would be the correct
compressed data to use.

However, I noticed that those tools produce compressed data that ends
the same. By using the last kilobyte of compressed data from 7-zip
(which was the same with Info-ZIP or WinZip), the attack was successful.
Found the keys: deadbeef baadf00d c0cac01a. I did not spend much effort
searching for a corresponding password. I can only tell it is at least
13 characters long. Anyways, the password is not needed to access data.
The following command creates an encrypted archive "images.zip"
containing the same files as "cipher.zip", but with the password of your
choice, "enjoy" for example.

bkcrack.exe -C cipher.zip -k deadbeef baadf00d c0cac01a -U images.zip enjoy

Then, you can decompress "images.zip" using any archive manager and the
password "enjoy".

For the record, afterwards I could decipher and check how the ciphertext
was compressed, and it does not start like any of the attempts I made
with various tools and settings. Good thing I used the end and not the
beginning!
~~~~

**Comment by kimci86** [comment]
It was solved via email. Glad I could help!
I investigated a little more using [infgen](https://github.com/madler/infgen). I found the deflated data starts with a dynamic block (it can be difficult to guess the right parameters) but ends with a stored block (not compressed). This is why it was easy to guess the end of compressed data: it is not compressed.
The takeaway from this case is that it can be worth using the end of compressed data instead of the beginning.
